src/Http/Http/src/StreamPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+
+                    var length = await _readingStream.ReadAsync(arraySegment.Array, arraySegment.Offset, arraySegment.Count, tokenSource.Token);
+#else
+#error Target frameworks need to be updated.
+#endif
+                    Debug.Assert(length + _readTail.End <= _readTail.AvailableMemory.Length);
+
+                    _readTail.End += length;
+                    _bufferedBytes += length;
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearOutCancellation();
+
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        throw;
+                    }
+
+                    isCanceled = true;
+                }
+
+                return new ReadResult(GetCurrentReadOnlySequence(), isCanceled, IsCompletedOrThrow());
+            }
+        }
+
+        private void ClearOutCancellation() => Interlocked.Exchange(ref _internalTokenSource, null);
+
+        private void ThrowIfCompleted()
+        {
+            if (_isCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+        }
+
+        public override bool TryRead(out ReadResult result)
+        {
+            ThrowIfCompleted();
+
+            return TryReadInternal(InternalTokenSource, out result);
+        }
+
+        private bool TryReadInternal(CancellationTokenSource source, out ReadResult result)
+        {
+            var isCancellationRequested = source.IsCancellationRequested;
+            if (isCancellationRequested || _bufferedBytes > 0 && !_examinedEverything)
+            {
+                // If TryRead/ReadAsync are called and cancellation is requested, we need to make sure memory is allocated for the ReadResult,
+                // otherwise if someone calls advance afterward on the ReadResult, it will throw.
+                if (isCancellationRequested)
+                {
+                    AllocateReadTail();
+
+                    ClearOutCancellation();
+                }
+
+                result = new ReadResult(
+                    GetCurrentReadOnlySequence(),
+                    isCanceled: isCancellationRequested,
+                    IsCompletedOrThrow());
+                return true;
+            }
+
+            result = new ReadResult();
+            return false;
+        }
+
+        private ReadOnlySequence<byte> GetCurrentReadOnlySequence()
+        {
+            return new ReadOnlySequence<byte>(_readHead, _readIndex, _readTail, _readTail.End - _readTail.Start);
+        }
+
+        private void AllocateReadTail()
+        {
+            if (_readHead == null)
+            {
+                Debug.Assert(_readTail == null);
+                _readHead = CreateSegmentUnsynchronized();
+                _readHead.SetMemory(_pool.Rent(GetSegmentSize()));
+                _readTail = _readHead;
+            }
+            else if (_readTail.WritableBytes == 0)
```

Comment:
Maybe create a new tail segment if WritableBytes is under a certain threshold. Kestrel transports will lease a new block to read into if there's less than 2KB of space in the existing tail segment.

Reply:
It’ll be added after this PR, we discussed this earlier